skactiveml/pool/_badge.py

```diff
@@ -0,0 +1,232 @@
+        else:
+            probas = clf.predict_proba(X_unlbld)
+        p_max = np.max(probas, axis=1).reshape(-1, 1)
+        g_x = (p_max - 1) * X_unlbld
+
+        # init the utilities
+        if mapping is not None:
+            utilities = np.full(
+                shape=(batch_size, X.shape[0]), fill_value=np.nan
+            )
+        else:
+            utilities = np.full(
+                shape=(batch_size, X_cand.shape[0]), fill_value=np.nan
+            )
+
+        # sampling with kmeans++
+        query_indicies = np.array([], dtype=int)
+        query_indicies_in_unlbld = np.array([], dtype=int)
+        d_2_s = []
+        for i in range(batch_size):
+            if i == 0:
+                d_2 = _d_2(g_x, [])
+            else:
+                d_2 = _d_2(g_x, [idx_in_unlbld], d_2_s[i - 1])
+            d_2_s.append(d_2)
+
+            d_2_sum = np.sum(d_2)
+            if i == 0 or d_2_sum == 0:
+                if d_2_sum == 0:
+                    d_2_s[-1] = np.full(shape=len(g_x), fill_value=np.inf)
+                d_2 = np.ones(shape=len(g_x))
+                d_2[query_indicies_in_unlbld] = 0
+                d_2_sum = np.sum(d_2)
+
+            d_probas = d_2 / d_2_sum
+
+            utilities[i, unlbld_mapping] = d_probas
+            utilities[i, query_indicies] = np.nan
+
+            idx_in_unlbld_array = self.random_state_.choice(
+                len(d_probas), 1, replace=False, p=d_probas
+            )
+            idx_in_unlbld = idx_in_unlbld_array[0]
+            query_indicies_in_unlbld = np.append(
+                query_indicies_in_unlbld, idx_in_unlbld_array
+            )
+
+            idx = unlbld_mapping[idx_in_unlbld]
+            query_indicies = np.append(query_indicies, [idx])
+
+        if return_utilities:
+            return query_indicies, utilities
+        else:
+            return query_indicies
+
+
+def _d_2(g_x, query_indicies, d_latest=None):
+    """
+    Calculates the D^2 value of the embedding features of unlabeled
+    data.
+
+    Parameters
+    ----------
+    g_x: numpy.ndarray of shape (n_unlabeled_samples, n_features)
+        The results after gradient embedding
+    query_indicies: numpy.ndarray of shape (n_query_indicies)
+        the query indicies, which mapping in the unlabeled samples
+    d_latest: numpy.ndarray of shape (n_unlabeled_samples) default=None
+        The distance between each data point and its nearest center
+        Using to facilitate the computation the later distances for the
+        coming selected sample
+
+    Returns
+    -------
+    D2: numpy.ndarray of shape (n_unlabeled_samples)
+        The D^2 value, for the first sample is the value inf.
+    """
+    if len(query_indicies) == 0:
+        return np.full(shape=len(g_x), fill_value=np.inf)
+    g_query_indicies = g_x[query_indicies]
+    _, D = pairwise_distances_argmin_min(X=g_x, Y=g_query_indicies)
+    if d_latest is not None:
+        D = np.minimum(d_latest, D)
+    D2 = np.square(D)
```

Comment:
```suggestion
    if d_latest is not None:
        D = np.minimum(d_latest, np.square(D))
    else:
        D2 = np.square(D)
```
If we always square the distances, it might happen that the `latest_distance` are squared multiple times.